src/ORM/DataObject.php

```diff
@@ -3175,14 +3175,24 @@ public static function reset()
     /**
      * Return the given element, searching by ID
      *
-     * @param string $callerClass The class of the object to be returned
+     * @param string $callerClass The class of the object to be returned (optional if called from extending class)
      * @param int $id The id of the element
      * @param boolean $cache See {@link get_one()}
      *
      * @return DataObject The element
      */
-    public static function get_by_id($callerClass, $id, $cache = true)
+    public static function get_by_id($callerClass, $id = null, $cache = true)
     {
+        if(is_numeric($callerClass)) {
+            // Alternate signature to allow calling as MyDataObject::get_by_id($id)
+            $cache = is_null($id) ? true : !!$id;
+            $id = $callerClass;
+            $callerClass = get_called_class();
+            if (__CLASS__ === $callerClass) {
+                user_error("DataObject::get_by_id called without providing object class", E_USER_WARNING);
```

Comment:
`InvalidArgumentException` is preferred.

Reply:
An Exception changes the signature because we'd have a @throws, but in practice it would still be a backwards compatible change because this position in the code can never be reached in any previously _working_ code (as no working code would ever send a numeric value as the first parameter).

Reply:
We don't declare `@throws` in silverstripe. We could, but don't.

GIven that there's no way this error could be thrown unless you use the NEW form of the method it doesn't feel like a semver violation to me.

Also, it's treated as a warning, but it's a real error. Not sure a warning is sufficient in any case.